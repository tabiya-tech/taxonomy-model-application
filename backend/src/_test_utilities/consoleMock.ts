/* eslint-disable @typescript-eslint/no-empty-function */
export {};
// Suppress chatty console during the tests
jest.spyOn(console, "error").mockImplementation(() => {});
jest.spyOn(console, "warn").mockImplementation(() => {});
jest.spyOn(console, "log").mockImplementation(() => {});
jest.spyOn(console, "info").mockImplementation(() => {});
