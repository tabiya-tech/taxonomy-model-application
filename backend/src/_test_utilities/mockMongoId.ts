import mongoose from "mongoose";

export function getMockStringId(index: number): string {
  return index.toString(16).padStart(24, "0");
}

export function getMockObjectId(index: number): mongoose.Types.ObjectId {
  return mongoose.Types.ObjectId.createFromHexString(getMockStringId(index));
}
