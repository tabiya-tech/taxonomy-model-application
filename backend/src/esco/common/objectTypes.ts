/**
 * Enum for the different types of objects in the ESCO ontology.
 */
export enum ObjectTypes {
  ISCOGroup = "ISCOGroup",
  Occupation = "Occupation",
  Skill = "Skill",
  SkillGroup = "SkillGroup",
}

/**
 * Enum for the two different types of relations in the ESCO ontology, essential and optional
 */
export enum RelationType {
  ESSENTIAL = "essential",
  OPTIONAL = "optional",
}

/**
 * Describes the three Types of Occupations
 */

export enum OccupationType {
  ESCO = "ESCO",
  LOCAL = "LOCAL",
  LOCALIZED = "LOCALIZED",
}

export type ReferenceWithRelationType<T> = T & {
  relationType: RelationType;
};

export type ImportIdentifiable = { importId: string };
