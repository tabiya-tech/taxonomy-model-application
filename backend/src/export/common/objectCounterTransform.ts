import { Transform, TransformCallback } from "stream";

export class ObjectCounterTransform extends Transform {
  private count: number = 0;

  constructor() {
    super({ objectMode: true });
  }

  _transform(chunk: unknown, _encoding: BufferEncoding, callback: TransformCallback): void {
    try {
      this.count++;
      this.push(chunk);
      callback(); // Always call callback after processing a chunk
    } catch (e: unknown) {
      // Proper error handling for streams
      callback(e as Error);
    }
  }

  getObjectCount(): number {
    return this.count;
  }

  _flush(callback: TransformCallback): void {
    callback();
  }
}
