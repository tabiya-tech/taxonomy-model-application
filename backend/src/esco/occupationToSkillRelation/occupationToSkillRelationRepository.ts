import mongoose from "mongoose";
import { ISkillDoc } from "esco/skill/skills.types";

import { ObjectTypes } from "esco/common/objectTypes";
import { isNewOccupationToSkillRelationPairSpecValid } from "./occupationToSkillRelationValidation";
import {
  INewOccupationToSkillPairSpec,
  IOccupationToSkillRelationPair,
  IOccupationToSkillRelationPairDoc,
} from "./occupationToSkillRelation.types";
import { handleInsertManyError } from "esco/common/handleInsertManyErrors";
import { IOccupationDoc } from "esco/occupation/occupation.types";
import { getModelName } from "esco/common/mongooseModelNames";

export interface IOccupationToSkillRelationRepository {
  readonly relationModel: mongoose.Model<IOccupationToSkillRelationPairDoc>;
  readonly skillModel: mongoose.Model<ISkillDoc>;
  readonly occupationModel: mongoose.Model<IOccupationDoc>;

  /**
   * Creates multiple new OccupationToSkillRelation entries.
   *
   * @param {modelId} modelId - The modelId of the model the relations belong to.
   * @param {INewOccupationToSkillPairSpec[]} newOccupationToSkillRelationPairSpecs - An array of specifications for the new OccupationToSkillRelation entries.
   * @return {Promise<IOccupationToSkillRelationPair[]>} - A Promise that resolves to an array containing the newly created OccupationToSkillRelation entries.
   * Excludes entries that fail validation and returns a subset of successfully created entries.
   * Rejects with an error if any entry cannot be created due to reasons other than validation.
   */
  createMany(
    modelId: string,
    newOccupationToSkillRelationPairSpecs: INewOccupationToSkillPairSpec[]
  ): Promise<IOccupationToSkillRelationPair[]>;
}

export class OccupationToSkillRelationRepository implements IOccupationToSkillRelationRepository {
  public readonly relationModel: mongoose.Model<IOccupationToSkillRelationPairDoc>;
  public readonly skillModel: mongoose.Model<ISkillDoc>;
  public readonly occupationModel: mongoose.Model<IOccupationDoc>;

  constructor(
    relationModel: mongoose.Model<IOccupationToSkillRelationPairDoc>,
    skillModel: mongoose.Model<ISkillDoc>,
    occupationModel: mongoose.Model<IOccupationDoc>
  ) {
    this.relationModel = relationModel;
    this.skillModel = skillModel;
    this.occupationModel = occupationModel;
  }

  async createMany(
    modelId: string,
    newOccupationToSkillRelationPairSpecs: INewOccupationToSkillPairSpec[]
  ): Promise<IOccupationToSkillRelationPair[]> {
    if (!mongoose.Types.ObjectId.isValid(modelId)) {
      const err = new Error(`Invalid modelId: ${modelId}`);
      console.error("batch create failed", err);
      throw err;
    }
    try {
      const existingIds = new Map<string, ObjectTypes>();

      // Get all Skills
      const _existingSkillsIds = await this.skillModel
        .find({ modelId: { $eq: modelId } })
        .select("_id")
        .exec();
      _existingSkillsIds.forEach((skill) => existingIds.set(skill._id.toString(), ObjectTypes.Skill));

      // Get all esco occupations
      const _existingOccupationsIds = await this.occupationModel
        .find({ modelId: { $eq: modelId } })
        .select("_id")
        .exec();
      _existingOccupationsIds.forEach((occupation) =>
        existingIds.set(occupation._id.toString(), ObjectTypes.Occupation)
      );
      // get all local occupations
      // get all localized occupations

      const newOccupationToSkillRelationPairModels = newOccupationToSkillRelationPairSpecs
        .filter((spec) => isNewOccupationToSkillRelationPairSpecValid(spec, existingIds))
        .map((spec) => {
          try {
            return new this.relationModel({
              ...spec,
              modelId: modelId,
              requiringOccupationDocModel: getModelName(spec.requiringOccupationType),
              requiredSkillDocModel: this.skillModel.modelName,
            });
          } catch (e: unknown) {
            return null;
          }
        })
        .filter(Boolean);

      const newRelations = await this.relationModel.insertMany(newOccupationToSkillRelationPairModels, {
        ordered: false,
      });
      if (newOccupationToSkillRelationPairSpecs.length !== newRelations.length) {
        console.warn(
          `OccupationToSkillRelationRepository.createMany: ${
            newOccupationToSkillRelationPairSpecs.length - newRelations.length
          } invalid entries were not created`
        );
      }
      return newRelations.map((pair) => pair.toObject());
    } catch (e: unknown) {
      return handleInsertManyError<IOccupationToSkillRelationPair>(
        e,
        "OccupationToSKillRelationRepository.createMany",
        newOccupationToSkillRelationPairSpecs.length
      );
    }
  }
}
