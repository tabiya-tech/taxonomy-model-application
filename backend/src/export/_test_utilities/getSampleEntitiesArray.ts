import { INewOccupationSpec, IOccupation } from "esco/occupation/occupation.types";
import { getMockRandomISCOGroupCode } from "_test_utilities/mockISCOCode";
import { getMockRandomOccupationCode } from "_test_utilities/mockOccupationCode";
import { ObjectTypes, OccupationType, RelationType } from "esco/common/objectTypes";
import { INewLocalizedOccupationSpec } from "esco/localizedOccupation/localizedOccupation.types";
import { INewISCOGroupSpec } from "esco/iscoGroup/ISCOGroup.types";
import { INewSkillSpec, ISkill, ReuseLevel, SkillType } from "esco/skill/skills.types";
import { INewSkillGroupSpec, ISkillGroup } from "esco/skillGroup/skillGroup.types";
import { getMockRandomSkillCode } from "_test_utilities/mockSkillGroupCode";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { INewOccupationHierarchyPairSpec } from "esco/occupationHierarchy/occupationHierarchy.types";
import { INewSkillHierarchyPairSpec } from "esco/skillHierarchy/skillHierarchy.types";
import { INewOccupationToSkillPairSpec } from "esco/occupationToSkillRelation/occupationToSkillRelation.types";
import { INewSkillToSkillPairSpec } from "esco/skillToSkillRelation/skillToSkillRelation.types";

export const getSampleOccupationSpecs = (givenModelId: string, isLocal: boolean = false): INewOccupationSpec[] => {
  return Array.from<never, INewOccupationSpec>({ length: 100 }, (_, i) => ({
    ISCOGroupCode: getMockRandomISCOGroupCode(),
    definition: `definition_${i}`,
    regulatedProfessionNote: `regulatedProfessionNote_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    code: getMockRandomOccupationCode(isLocal),
    preferredLabel: `Occupation_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: isLocal ? OccupationType.LOCAL : OccupationType.ESCO,
  }));
};
export const getSampleLocalizedOccupationSpecs = (
  givenEscoOccupations: IOccupation[]
): INewLocalizedOccupationSpec[] => {
  return givenEscoOccupations.map((occupation, i) => ({
    localizesOccupationId: occupation.id,
    altLabels: i % 2 ? [] : [`altLabel_1`, `altLabel_2`],
    modelId: occupation.modelId,
    description: `description_${i}`,
    importId: `importId_${i}`,
    occupationType: OccupationType.LOCALIZED,
  }));
};

export const getSampleISCOGroupSpecs = (givenModelId: string): INewISCOGroupSpec[] => {
  return Array.from<never, INewISCOGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomISCOGroupCode(),
    preferredLabel: `ISCOGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsSpecs = (givenModelId: string): INewSkillSpec[] => {
  return Array.from<never, INewSkillSpec>({ length: 100 }, (_, i) => ({
    preferredLabel: `Skill_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    definition: `definition_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    skillType: SkillType.Knowledge,
    reuseLevel: ReuseLevel.CrossSector,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillGroupsSpecs = (givenModelId: string): INewSkillGroupSpec[] => {
  return Array.from<never, INewSkillGroupSpec>({ length: 100 }, (_, i) => ({
    code: getMockRandomSkillCode(),
    preferredLabel: `SkillGroup_${i}`,
    modelId: givenModelId,
    originUUID: ``,
    ESCOUri: `ESCOUri_${i}`,
    description: `description_${i}`,
    scopeNote: `scopeNote_${i}`,
    altLabels: i % 2 ? [`altLabel_1`, `altLabel_2`] : [],
    importId: `importId_${i}`,
  }));
};

export const getSampleOccupationHierarchy = (givenModelId: string): INewOccupationHierarchyPairSpec[] => {
  return Array.from<never, INewOccupationHierarchyPairSpec>({ length: 100 }, (_, i) => ({
    parentId: getMockStringId(999),
    parentType: ObjectTypes.Occupation,
    childType: ObjectTypes.Occupation,
    childId: getMockStringId(i),
    modelId: givenModelId,
    importId: `importId_${i}`,
  }));
};

export const getSampleSkillsHierarchy = (
  givenSkills: ISkill[],
  givenSkillGroups: ISkillGroup[]
): INewSkillHierarchyPairSpec[] => {
  return givenSkills.map((skill) => ({
    parentId: givenSkillGroups[0].id,
    parentType: ObjectTypes.SkillGroup,
    childType: ObjectTypes.Skill,
    childId: skill.id,
    modelId: skill.modelId,
  }));
};

export const getSampleOccupationToSkillRelations = (
  givenOccupations: IOccupation[],
  givenSkills: ISkill[]
): INewOccupationToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringOccupationType: OccupationType.LOCAL,
    requiredSkillId: skill.id,
    requiringOccupationId: givenOccupations[i].id,
    relationType: RelationType.OPTIONAL,
  }));
};

export const getSampleSkillToSkillRelations = (givenSkills: ISkill[]): INewSkillToSkillPairSpec[] => {
  return givenSkills.map((skill, i) => ({
    requiringSkillId: givenSkills[i + 1]?.id ?? givenSkills[0].id,
    requiredSkillId: skill.id,
    relationType: RelationType.OPTIONAL,
  }));
};
