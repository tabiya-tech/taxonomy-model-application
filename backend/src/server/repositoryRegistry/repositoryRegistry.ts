import { IModelRepository, ModelRepository } from "modelInfo/modelInfoRepository";
import { Connection } from "mongoose";
import * as modelInfoModel from "modelInfo/modelInfoModel";
import * as ISCOGroupModel from "esco/iscoGroup/ISCOGroupModel";
import * as skillGroupModel from "esco/skillGroup/skillGroupModel";
import * as skillModel from "esco/skill/skillModel";
import * as occupationModel from "esco/occupation/occupationModel";
import * as localizedOccupationModel from "esco/localizedOccupation/localizedOccupationModel";
import * as occupationHierarchyModel from "esco/occupationHierarchy/occupationHierarchyModel";
import * as skillHierarchyModel from "esco/skillHierarchy/skillHierarchyModel";
import * as skillToSkillRelationModel from "esco/skillToSkillRelation/skillToSkillRelationModel";
import * as occupationToSkillRelationModel from "esco/occupationToSkillRelation/occupationToSkillRelationModel";
import * as importStateModel from "import/ImportProcessState/importProcessStateModel";
import * as exportProcessStateModel from "export/exportProcessState/exportProcessStateModel";

import { IISCOGroupRepository, ISCOGroupRepository } from "esco/iscoGroup/ISCOGroupRepository";
import { ISkillGroupRepository, SkillGroupRepository } from "esco/skillGroup/skillGroupRepository";
import { ISkillRepository, SkillRepository } from "esco/skill/skillRepository";
import { IOccupationRepository, OccupationRepository } from "esco/occupation/occupationRepository";
import {
  IOccupationHierarchyRepository,
  OccupationHierarchyRepository,
} from "esco/occupationHierarchy/occupationHierarchyRepository";
import {
  IImportProcessStateRepository,
  ImportProcessStateRepository,
} from "import/ImportProcessState/importProcessStateRepository";
import { ISkillHierarchyRepository, SkillHierarchyRepository } from "esco/skillHierarchy/skillHierarchyRepository";
import {
  ISkillToSkillRelationRepository,
  SkillToSkillRelationRepository,
} from "esco/skillToSkillRelation/skillToSkillRelationRepository";
import {
  IOccupationToSkillRelationRepository,
  OccupationToSkillRelationRepository,
} from "esco/occupationToSkillRelation/occupationToSkillRelationRepository";
import {
  ILocalizedOccupationRepository,
  LocalizedOccupationRepository,
} from "esco/localizedOccupation/localizedOccupationRepository";
import {
  ExportProcessStateRepository,
  IExportProcessStateRepository,
} from "export/exportProcessState/exportProcessStateRepository";

export class RepositoryRegistry {
  // eslint-disable-next-line
  private readonly _repositories: Map<string, any> = new Map<string, any>();

  public get modelInfo(): IModelRepository {
    return this._repositories.get("ModelRepository");
  }

  public set modelInfo(repository: IModelRepository) {
    this._repositories.set("ModelRepository", repository);
  }

  public get ISCOGroup(): IISCOGroupRepository {
    return this._repositories.get("ISCOGroupRepository");
  }

  public set ISCOGroup(repository: IISCOGroupRepository) {
    this._repositories.set("ISCOGroupRepository", repository);
  }

  public get skillGroup(): ISkillGroupRepository {
    return this._repositories.get("ISkillGroupRepository");
  }

  public set skillGroup(repository: ISkillGroupRepository) {
    this._repositories.set("ISkillGroupRepository", repository);
  }

  public get skill(): ISkillRepository {
    return this._repositories.get("ISkillRepository");
  }

  public set skill(repository: ISkillRepository) {
    this._repositories.set("ISkillRepository", repository);
  }

  public get occupation(): IOccupationRepository {
    return this._repositories.get("IOccupationRepository");
  }

  public set occupation(repository: IOccupationRepository) {
    this._repositories.set("IOccupationRepository", repository);
  }

  public get localizedOccupation(): ILocalizedOccupationRepository {
    return this._repositories.get("ILocalizedOccupationRepository");
  }

  public set localizedOccupation(repository: ILocalizedOccupationRepository) {
    this._repositories.set("ILocalizedOccupationRepository", repository);
  }

  public get occupationHierarchy(): IOccupationHierarchyRepository {
    return this._repositories.get("IOccupationHierarchyRepository");
  }

  public set occupationHierarchy(repository: IOccupationHierarchyRepository) {
    this._repositories.set("IOccupationHierarchyRepository", repository);
  }

  public get skillHierarchy(): ISkillHierarchyRepository {
    return this._repositories.get("ISkillHierarchyRepository");
  }

  public set skillHierarchy(repository: ISkillHierarchyRepository) {
    this._repositories.set("ISkillHierarchyRepository", repository);
  }

  public get skillToSkillRelation(): ISkillToSkillRelationRepository {
    return this._repositories.get("ISkillToSkillRelationRepository");
  }

  public set skillToSkillRelation(repository: ISkillToSkillRelationRepository) {
    this._repositories.set("ISkillToSkillRelationRepository", repository);
  }

  public set occupationToSkillRelation(repository: IOccupationToSkillRelationRepository) {
    this._repositories.set("IOccupationToSkillRelationRepository", repository);
  }

  public get occupationToSkillRelation(): IOccupationToSkillRelationRepository {
    return this._repositories.get("IOccupationToSkillRelationRepository");
  }

  public get importProcessState(): IImportProcessStateRepository {
    return this._repositories.get("IImportProcessStateRepository");
  }

  public set importProcessState(repository: IImportProcessStateRepository) {
    this._repositories.set("IImportProcessStateRepository", repository);
  }

  public get exportProcessState(): IExportProcessStateRepository {
    return this._repositories.get("IExportProcessStateRepository");
  }

  public set exportProcessState(repository: IExportProcessStateRepository) {
    this._repositories.set("IExportProcessStateRepository", repository);
  }

  async initialize(connection: Connection | undefined) {
    if (!connection) throw new Error("Connection is undefined");

    // Set up the ModelRepository
    this.modelInfo = new ModelRepository(modelInfoModel.initializeSchemaAndModel(connection));
    this.ISCOGroup = new ISCOGroupRepository(ISCOGroupModel.initializeSchemaAndModel(connection));
    this.skillGroup = new SkillGroupRepository(skillGroupModel.initializeSchemaAndModel(connection));
    this.skill = new SkillRepository(skillModel.initializeSchemaAndModel(connection));
    this.occupation = new OccupationRepository(occupationModel.initializeSchemaAndModel(connection));
    this.localizedOccupation = new LocalizedOccupationRepository(
      localizedOccupationModel.initializeSchemaAndModel(connection),
      this.occupation.Model
    );
    this.occupationHierarchy = new OccupationHierarchyRepository(
      occupationHierarchyModel.initializeSchemaAndModel(connection),
      this.ISCOGroup.Model,
      this.occupation.Model
    );
    this.skillHierarchy = new SkillHierarchyRepository(
      skillHierarchyModel.initializeSchemaAndModel(connection),
      this.skill.Model,
      this.skillGroup.Model
    );
    this.skillToSkillRelation = new SkillToSkillRelationRepository(
      skillToSkillRelationModel.initializeSchemaAndModel(connection),
      this.skill.Model
    );
    this.occupationToSkillRelation = new OccupationToSkillRelationRepository(
      occupationToSkillRelationModel.initializeSchemaAndModel(connection),
      this.skill.Model,
      this.occupation.Model,
      this.localizedOccupation.Model
    );
    this.importProcessState = new ImportProcessStateRepository(importStateModel.initializeSchemaAndModel(connection));
    this.exportProcessState = new ExportProcessStateRepository(
      exportProcessStateModel.initializeSchemaAndModel(connection)
    );

    // Set up the indexes
    // This is done here because the autoIndex is turned off in production
    // In a production environment,
    // the indexes must be created manually before the application is started for the first time,
    // and it the future this code should be moved in to deployment scripts.
    // If indexes are not created then, queries will become inefficient, unique constrains will not be enforced.
    await this.modelInfo.Model.createIndexes();
    await this.ISCOGroup.Model.createIndexes();
    await this.skillGroup.Model.createIndexes();
    await this.skill.Model.createIndexes();
    await this.occupation.Model.createIndexes();
    await this.localizedOccupation.Model.createIndexes();
    await this.occupationHierarchy.hierarchyModel.createIndexes();
    await this.skillHierarchy.hierarchyModel.createIndexes();
    await this.skillToSkillRelation.relationModel.createIndexes();
    await this.occupationToSkillRelation.relationModel.createIndexes();
    await this.importProcessState.Model.createIndexes();
    await this.exportProcessState.Model.createIndexes();
  }
}

const _repositoryRegistryInstance: RepositoryRegistry = new RepositoryRegistry();

export function getRepositoryRegistry(): RepositoryRegistry {
  return _repositoryRegistryInstance;
}
