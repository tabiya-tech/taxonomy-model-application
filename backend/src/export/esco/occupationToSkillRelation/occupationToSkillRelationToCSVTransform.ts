import { IOccupationToSkillRelationPair } from "esco/occupationToSkillRelation/occupationToSkillRelation.types";
import {
  IOccupationToSkillRelationExportRow,
  occupationToSkillRelationExportHeaders,
} from "esco/common/entityToCSV.types";
import { pipeline, Transform } from "stream";
import { stringify } from "csv-stringify";
import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { Readable } from "node:stream";
import {
  getCSVTypeFromObjectType,
  getCSVRelationTypeFromOccupationToSkillRelationType,
} from "esco/common/csvObjectTypes";

export type IUnpopulatedOccupationToSkillRelation = Omit<
  IOccupationToSkillRelationPair,
  "requiringOccupationDocModel" | "requiredSkillDocModel"
>;

export const transformOccupationToSkillRelationSpecToCSVRow = (
  occupationToSkillRelation: IUnpopulatedOccupationToSkillRelation
): IOccupationToSkillRelationExportRow => {
  const OCCUPATIONTYPE = getCSVTypeFromObjectType(occupationToSkillRelation.requiringOccupationType);
  if (OCCUPATIONTYPE === null) {
    throw new Error(
      `Failed to transform OccupationToSkillRelation to CSV row: Invalid requiringOccupationType: ${occupationToSkillRelation.requiringOccupationType}`
    );
  }
  const RELATIONTYPE = getCSVRelationTypeFromOccupationToSkillRelationType(occupationToSkillRelation.relationType);
  if (RELATIONTYPE === null) {
    throw new Error(
      `Failed to transform OccupationToSkillRelation to CSV row: Invalid relationType: ${occupationToSkillRelation.relationType}`
    );
  }
  return {
    //@ts-ignore
    OCCUPATIONTYPE,
    OCCUPATIONID: occupationToSkillRelation.requiringOccupationId,
    RELATIONTYPE,
    SKILLID: occupationToSkillRelation.requiredSkillId,
    CREATEDAT: occupationToSkillRelation.createdAt.toISOString(),
    UPDATEDAT: occupationToSkillRelation.updatedAt.toISOString(),
  };
};

class OccupationToSkillRelationToCSVRowTransformer extends Transform {
  constructor() {
    super({ objectMode: true });
  }

  _transform(
    occupationToSkillRelation: IUnpopulatedOccupationToSkillRelation,
    _encoding: BufferEncoding,
    callback: (error?: Error | null, data?: never) => void
  ): void {
    try {
      const transformedRow = transformOccupationToSkillRelationSpecToCSVRow(occupationToSkillRelation);
      this.push(transformedRow);
      callback();
    } catch (cause: unknown) {
      // Make sure stringification doesn't fail, otherwise throwing an error will cause the stream to hang
      let json: string = "";
      try {
        json = JSON.stringify(occupationToSkillRelation, null, 2);
      } finally {
        const err = new Error(`Failed to transform occupationToSkillRelation to CSV row: ${json}`, { cause: cause });
        console.error(err);
        callback(err);
      }
    }
  }
}

const OccupationToSkillRelationToCSVTransform = (modelId: string): Readable => {
  // the stringify is a stream, and we need a new one every time we create a new pipeline
  const occupationToSkillRelationStringifier = stringify({
    header: true,
    columns: occupationToSkillRelationExportHeaders,
  });

  return pipeline(
    getRepositoryRegistry().occupationToSkillRelation.findAll(modelId),
    new OccupationToSkillRelationToCSVRowTransformer(),
    occupationToSkillRelationStringifier,
    (cause) => {
      if (cause) {
        console.error(new Error("Transforming occupationToSkillRelation to CSV failed", { cause: cause }));
      }
    }
  );
};

export default OccupationToSkillRelationToCSVTransform;
