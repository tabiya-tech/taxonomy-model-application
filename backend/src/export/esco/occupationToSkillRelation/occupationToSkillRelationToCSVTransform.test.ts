// Mute chatty console logs
import "_test_utilities/consoleMock";

import { getRepositoryRegistry } from "server/repositoryRegistry/repositoryRegistry";
import { getMockStringId } from "_test_utilities/mockMongoId";
import { OccupationType, RelationType } from "esco/common/objectTypes";
import { Readable } from "node:stream";
import occupationToSkillRelationToCSVTransform, {
  IUnpopulatedOccupationToSkillRelation,
} from "./occupationToSkillRelationToCSVTransform";
import { IOccupationToSkillRelationRepository } from "esco/occupationToSkillRelation/occupationToSkillRelationRepository";
import * as occupationToSkillRelationToCSVTransformModule from "./occupationToSkillRelationToCSVTransform";
import { parse } from "csv-parse/sync";

const OccupationToSkillRelationRepositorySpy = jest.spyOn(getRepositoryRegistry(), "occupationToSkillRelation", "get");

const getMockOccupationToSkillRelations = (): IUnpopulatedOccupationToSkillRelation[] => {
  function getOccupationType(i: number) {
    switch (i % 3) {
      case 0:
        return OccupationType.ESCO;
      case 1:
        return OccupationType.LOCAL;
      case 2:
        return OccupationType.LOCALIZED;
      default:
        throw new Error("Invalid number");
    }
  }

  return Array.from({ length: 6 }, (_, i) => ({
    id: getMockStringId(i * 3),
    modelId: getMockStringId(1),
    requiringOccupationId: getMockStringId(i * 3 + 1),
    requiredSkillId: getMockStringId(i * 3 + 2),
    requiringOccupationType: getOccupationType(i),
    relationType: i % 2 ? RelationType.ESSENTIAL : RelationType.OPTIONAL,
    createdAt: new Date(i), // use a fixed date to make the snapshot stable
    updatedAt: new Date(i), // use a fixed date to make the snapshot stable
  }));
};

function setupOccupationToSkillRelationRepositoryMock(findAllImpl: () => Readable) {
  const mockOccupationToSkillRelationRepository: IOccupationToSkillRelationRepository = {
    relationModel: undefined as never,
    occupationModel: undefined as never,
    localizedOccupationModel: undefined as never,
    skillModel: undefined as never,
    createMany: jest.fn().mockResolvedValue(null),
    findAll: jest.fn().mockImplementationOnce(findAllImpl),
  };
  OccupationToSkillRelationRepositorySpy.mockReturnValue(mockOccupationToSkillRelationRepository);
}

describe("occupationToSkillRelationToCSVTransform", () => {
  test("should correctly transform occupationToSkillRelation data to CSV", async () => {
    // GIVEN findAll returns a stream of occupationToSkillRelations
    const givenRelations = getMockOccupationToSkillRelations();
    setupOccupationToSkillRelationRepositoryMock(() => Readable.from(givenRelations));

    // WHEN the transformation is applied
    const transformedStream = occupationToSkillRelationToCSVTransform("foo");

    // THEN expect the output to be a valid CSV
    const chunks = [];
    for await (const chunk of transformedStream) {
      chunks.push(chunk);
    }
    const actualCSVOutput = chunks.join("");
    const parsedObjects = parse(actualCSVOutput, { columns: true });
    // AND contain the given occupationToSkillRelation data
    expect(parsedObjects).toMatchSnapshot();
    expect(actualCSVOutput).toMatchSnapshot();
    // AND the stream should end
    expect(transformedStream.closed).toBe(true);
  });

  describe("handle errors during stream processing", () => {
    test("should log an error and end the stream when the source repository fails", async () => {
      // GIVEN that the source occupationToSkillRelation stream will emit an error
      const givenError = new Error("Test Error");
      setupOccupationToSkillRelationRepositoryMock(
        () =>
          new Readable({
            read() {
              this.emit("error", givenError);
            },
          })
      );

      // WHEN the transformation stream is consumed
      const transformedStream = occupationToSkillRelationToCSVTransform("foo");

      // THEN expect the given error to be thrown
      await expect(async () => {
        //  iterate to consume the stream
        for await (const _ of transformedStream) {
          // do nothing
        }
      }).rejects.toThrowError(givenError);
      // AND the error should be logged
      expect(console.error).toHaveBeenNthCalledWith(1, expect.any(Error), expect.any(Error));
      // AND the stream should end
      expect(transformedStream.closed).toBe(true);
    });

    test("should log an error and end the stream when the CSV stringifier fails", async () => {
      // GIVEN findAll returns a stream of occupationToSkillRelations
      setupOccupationToSkillRelationRepositoryMock(() => Readable.from(getMockOccupationToSkillRelations()));

      // AND  the transformOccupationToSkillRelationSpecToCSVRow will throw an error
      const givenError = new Error("Mocked Transformation Error");
      jest
        .spyOn(occupationToSkillRelationToCSVTransformModule, "transformOccupationToSkillRelationSpecToCSVRow")
        .mockImplementationOnce(
          () => {
            throw givenError;
          }
        );

      // WHEN the transformation stream is consumed
      const transformedStream = occupationToSkillRelationToCSVTransform("foo");

      // THEN expect the given error to be thrown
      await expect(async () => {
        //  iterate to consume the stream
        for await (const _ of transformedStream) {
          // do nothing
        }
      }).rejects.toThrowError("Failed to transform occupationToSkillRelation to CSV row");
      // AND the error to be logged
      expect(console.error).toHaveBeenNthCalledWith(1, expect.any(Error), expect.any(Error));
      expect(console.error).toHaveBeenNthCalledWith(2, expect.any(Error), expect.any(Error));
      // AND the stream to end
      expect(transformedStream.closed).toBe(true);
    });
  });
});
