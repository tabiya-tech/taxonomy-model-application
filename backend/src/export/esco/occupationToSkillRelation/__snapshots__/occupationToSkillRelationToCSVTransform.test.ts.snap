// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`occupationToSkillRelationToCSVTransform should correctly transform occupationToSkillRelation data to CSV 1`] = `
[
  {
    "OCCUPATIONID": "000000000000000000000001",
    "OCCUPATIONTYPE": "ESCO",
    "RELATIONTYPE": "optional",
    "SKILLID": "000000000000000000000002",
  },
  {
    "OCCUPATIONID": "000000000000000000000004",
    "OCCUPATIONTYPE": "LOCAL",
    "RELATIONTYPE": "essential",
    "SKILLID": "000000000000000000000005",
  },
  {
    "OCCUPATIONID": "000000000000000000000007",
    "OCCUPATIONTYPE": "LOCALIZED",
    "RELATIONTYPE": "optional",
    "SKILLID": "000000000000000000000008",
  },
  {
    "OCCUPATIONID": "00000000000000000000000a",
    "OCCUPATIONTYPE": "ESCO",
    "RELATIONTYPE": "essential",
    "SKILLID": "00000000000000000000000b",
  },
  {
    "OCCUPATIONID": "00000000000000000000000d",
    "OCCUPATIONTYPE": "LOCAL",
    "RELATIONTYPE": "optional",
    "SKILLID": "00000000000000000000000e",
  },
  {
    "OCCUPATIONID": "000000000000000000000010",
    "OCCUPATIONTYPE": "LOCALIZED",
    "RELATIONTYPE": "essential",
    "SKILLID": "000000000000000000000011",
  },
]
`;

exports[`occupationToSkillRelationToCSVTransform should correctly transform occupationToSkillRelation data to CSV 2`] = `
"OCCUPATIONTYPE,OCCUPATIONID,RELATIONTYPE,SKILLID
ESCO,000000000000000000000001,optional,000000000000000000000002
LOCAL,000000000000000000000004,essential,000000000000000000000005
LOCALIZED,000000000000000000000007,optional,000000000000000000000008
ESCO,00000000000000000000000a,essential,00000000000000000000000b
LOCAL,00000000000000000000000d,optional,00000000000000000000000e
LOCALIZED,000000000000000000000010,essential,000000000000000000000011
"
`;
